### ISSUE TITLE
Reducing ByteTensor in Variable gives wrong answer

### ISSUE THREAD

**Issue by awni** [pytorch/pytorch#2124]
When summing a `ByteTensor` wrapped in a `Variable` the answer is incorrect. The answer can be very off in some cases. See the below gist for reproducing.

https://gist.github.com/awni/675bedbda43e7aa3ad06dfd7f5ce7205

**Comment by soumith** [comment]
i believe this is because ByteTensor has limited precision, and can only represent 8-bit numbers. If your reduction overflows, you'll see these issues.

**Comment by awni** [comment]
I see..
It originally came up because I was comparing two `IntTensors` - `torch.sum(a==b)`. The `a==b` is a `ByteTensor` and the sum gives the wrong answer (but oddly only for the `Variable`, not for a plain Tensor).

It may be a feature request then to have the Variable sum behave the same as the Tensor / Numpy sum?

**Comment by lynic** [comment]
Is it better to return IntTensor instead of ByteTensor when doing '==' operate, so that when we do sum or other operations it won't easily overflow?

**Comment by awni** [comment]
I think it would be good to have `torch.sum` behave the same on a plain `ByteTensor` as it does on a `ByteTensor` which is wrapped in a `Variable`. Presumably for the plain `ByteTensor` the accumulation is done in higher precision whereas for the `Variable` it seems like it's overflowing.

If you keep the `ByteTensor` reduce in higher precision then I think it's fine to have `==` return a `ByteTensor` (this also seems more inline with numpy semantics which returns a bool on `a==b`).

**Comment by lantiga** [comment]
The thing is that, given
```
a = torch.ByteTensor([1,2])
```
`sum` returns a raw number if called without arguments (it sums all the elements of the tensor)
```
>>> type(torch.sum(a))
<type 'int'>
```
while it returns a new `ByteTensor` when `dim` is specified
```
>>> type(torch.sum(a,dim=0))
<type 'torch.ByteTensor'>
```
so it would overflow if `dim` is provided.

On the other hand, `sum` on a `Variable` always wraps the results in a `ByteTensor` variable, with our without `dim`.

Looking at the API I would say that `torch.sum(a)` is the exception here, although the `a==b` case is admittedly a big gotcha and probably a common use case.

**Comment by gchanan** [comment]
This has been fixed.
